IPython/sphinxext/ipython_directive.py

```diff
@@ -790,74 +790,56 @@ def process_pure_python(self, content):
         the content as a list as if it were ipython code
         """
         output = []
-        savefig = False # keep up with this to clear figure
-        multiline = False # to handle line continuation
-        multiline_start = None
         fmtin = self.promptin
-
         ct = 0
 
-        for lineno, line in enumerate(content):
-
-            line_stripped = line.strip()
-            if not len(line):
-                output.append(line)
-                continue
-
-            # handle decorators
-            if line_stripped.startswith('@'):
-                output.extend([line])
-                if 'savefig' in line:
-                    savefig = True # and need to clear figure
+        try:
+            savefigs = [i for i, line in enumerate(content)
+                        if line[:8] == '@savefig']
+            comments = [i for i, line in enumerate(content)
+                        if line != '' and line[0] == '#']
+            content_text = '\n'.join([c if c[:8] != '@savefig' else ''
+                                      for c in content])
+            mod = ast.parse(content_text)
+            linestarts = [body.lineno - 1 for body in mod.body]
+        except Exception:
+            savefigs = []
+            comments = []
+            linestarts = [0]
+
+        breaks = sorted(savefigs + linestarts + comments)
+        for i, lineno in enumerate(breaks):
+            if i == 0:
                 continue
+            output.append([c for c in content[breaks[i-1]:lineno]])
+
+        output.append([c for c in content[lineno:]])
```

Comment:
what if there was no breaks ? will lineno be undefined ?

Reply:
Yeah I need to write some testcases for this.